### ISSUE TITLE
Rework Problem class

### ISSUE THREAD

**Issue by fKunstner** [fKunstner/optexp#5]
The problem class currently defines some pieces of the training loop. 

This is a but much, and some of it could be moved out of it and into the training loop.

The biggest problem with the Problem class is the hard-coded definition of the metrics, which we can get around if we introduce metrics separately as in #4. If we do this we can get rid of 90% of the code in this folder.

But I'd still want to keep the concept of a `Problem` around. 
From the perspective of evaluating optimizers, the `Problem` is what we'd like to keep constant and try many optimizers against it.
Having a consistent naming scheme and a set of problem would be nice for this, 
and we can't really get there by taking the cross product of datasets and models because some models only work with some datasets and don't make sense otherwise.  `Problem` seems to be a nice abstraction for that.

And that abstraction might actually be necessary if we need to introduce more complicated "problems", say training a GAN, 
that would have a different way to compute the loss function that is technically different from just `loss(m(x),y)`
